### PR TITLE
Refactor for fault tolerance

### PR DESCRIPTION
**Pull request by sonallux**
Currently, this library tries to hide all most of the details of the underlying `amqplib` and does expose new functions to either publish or consume AMQP messages. To make our event system in the ODS more fault-tolerant and robust we do need some functionality of the `amqplib` that is currently not exposed by `node-dry-amqp`:

- [consumer acknowledgement](https://www.rabbitmq.com/confirms.html#acknowledgement-modes): So successfully consumed messages are removed from the queue
- [publisher confirms](https://www.rabbitmq.com/confirms.html#publisher-confirms): So the publisher does get feedback if the message really has been published successfully. Currently, only the successful message transportation over the network to the AMQP broker is guaranteed if the publish message is returning. The successful routing to a queue is currently not guaranteed, but it can be with a publisher confirms.
- connection and channel error handling: If a network error happens the connection and channel are closed and can not be reused.

I think the `amqplib` does actually already provide a good API surface, the only things that are missing are connection retry and connection error handling. Therefore I would not expose a completely new API with this library and instead, just add a wrapper around `amqplib` that can perform connection retries and can survive network errors by recreating an underlying connection/channel if needed.

### New features
- The new `AmqpConnection`, `AmqpChannel` and `AmqpConfirmChannel` classes are delegates that are providing the same methods as their counterparts from `amqplib`. Just with the addition that they are recreating the underlying `amqplib` objects if needed. Therefore objects of the new classes can still be used after a connection failure.
- The `AmqpConnection` class can also be configured to perform retries when connecting to the AMQP broker.
- After a connection failure the `AmqpConnection` is automatically trying to reconnect to the AMQP broker. If that fails a callback is invoked so the application can decide how to handle this situation. This feature is mainly needed for the next point.
- The `AmqpChannel` and `AmqpConfirmChannel` do restore registered consumers after a connection failure has been resolved.

### CODE REVIEW

**Comment by sonallux** [comment]
This is ready for review. I have also updated the PR description, so please have a look there for details about the changes.

Happy do hear your feedback @georg-schwarz and @lunedis.

File: src/amqpChannel.ts [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+  consumer: AmqpConsumer
+  options?: AMQP.Options.Consume
+}
+
+/**
+ * Common super class for an amqp channel
+ */
+export class BaseAmqpChannel<T extends AMQP.Channel> {
+  protected rawChannel?: T = undefined
+  protected readonly consumers = new Map<string, StoredConsumer>()
```
Comment (georg-schwarz) [inline src/amqpChannel.ts:23]: so you can have many consumers, one per what?

File: src/amqpConnection.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+      setTimeout(() => {
+        this.getRawConnection().catch(this.connectionLossHandler)
+      }, this.retryDelayInMs)
+    })
+    // If the connection has been closed it can not be reused, so we can remove the reference
+    connection.on('close', () => {
+      this.rawConnection = undefined
+    })
+    this.rawConnection = connection
+    // Inform all registered channels about the successfully established connection
```
Comment (georg-schwarz) [inline src/amqpConnection.ts]: extract to method => makes comment unnecessary

File: src/amqpChannel.ts [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+    return await channel.prefetch(count, global)
+  }
+
+  async recover (): Promise<AMQP.Replies.Empty> {
+    const channel = await this.getRawChannel()
+    return await channel.recover()
+  }
+}
+
+export class AmqpChannel extends BaseAmqpChannel<AMQP.Channel> {
```
Comment (georg-schwarz) [inline src/amqpChannel.ts:190]: File got quite huge, can we pull sth into a new file?

File: src/amqpChannel.ts (hunk 1 above)
Comment (sonallux) [reply src/amqpChannel.ts:23]: Sorry, but I did not get your question. One can register multiple consumers per channel in AMQP. This map stores all registered consumers with the consumerTag as key.

File: src/amqpChannel.ts (hunk 3 above)
Comment (sonallux) [reply src/amqpChannel.ts:190]: Well, we could extract the `AmqpChannel` and `AmqpConfirmChannel` into a new file

Comment (sonallux) [reply src/amqpChannel.ts:190]: But that would only reduce the file by ~30 lines. So I do not know if that's really worth it.

File: src/amqpConnection.ts (hunk 2 above)
Comment (sonallux) [reply src/amqpConnection.ts]: Done

File: src/amqpChannel.ts (hunk 3 above)
Comment (georg-schwarz) [reply src/amqpChannel.ts:190]: Allright, then let's leave it that way!

File: src/amqpChannel.ts (hunk 1 above)
Comment (georg-schwarz) [reply src/amqpChannel.ts:23]: ah, so it is just an identifier for the consumer, I see.

**Review by georg-schwarz — APPROVED** [review]
LGTM

File: src/amqpChannel.ts (hunk 1 above)
Comment (sonallux) [reply src/amqpChannel.ts:23]: Yes exactly

**Review by lunedis — APPROVED** [review]
Looks good to me as well.

**Comment by georg-schwarz** [comment]
Minor release? Didn't see an existing API change, right?

**Comment by sonallux** [comment]
> Minor release? Didn't see an existing API change, right?

Yes, the new version should be `0.1.0`.

**Pull request merged**